### PR TITLE
CASSANDRA-19901: Refactor TokenRangeMapping to use proper types instead of Strings

### PR DESCRIPTION
**Pull request by yifan-c**
Patch by Yifan Cai; Reviewed by TBD for CASSANDRA-19901

### CODE REVIEW

File: scripts/build-dtest-jars.sh [hunk 1]
```diff
@@ -118,4 +118,6 @@ else
         exit ${RETURN}
     fi
   done
+  # always delete the Cassandra source after dtest.jar is built to avoid confusing IDE
```
Comment (frankgh) [inline scripts/build-dtest-jars.sh:121]: can we alternatively exclude these files from the IDE's indexing?

File: scripts/build-sidecar.sh [hunk 2]
```diff
@@ -77,4 +77,6 @@ else
   fi
   git clean -fd
   ./gradlew -Pversion=${SIDECAR_BUILD_VERSION} -Dmaven.repo.local=${SIDECAR_JAR_DIR} publishToMavenLocal
+  # Delete sidecar source after publishing to avoid confusing IDE
+  rm -rf "${SIDECAR_BUILD_DIR}"
```
Comment (frankgh) [inline scripts/build-sidecar.sh:81]: can we alternatively exclude these files from the IDE's indexing?

File: cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/bulkwriter/token/TokenRangeMapping.java [hunk 3]
```diff
@@ -22,55 +22,131 @@
...
+
+        if (LOGGER.isDebugEnabled())
+        {
+            LOGGER.debug("Fetched token-ranges with dcs={}, write_replica_count={}, pending_replica_count={}",
+                         writeReplicasByDC.keySet(),
+                         writeReplicasByDC.values().stream().flatMap(Collection::stream).collect(Collectors.toSet()).size(),
+                         pendingReplicasByDC.values().stream().flatMap(Collection::stream).collect(Collectors.toSet()).size());
+        }
+
+        Map<String, ReplicaMetadata> replicaMetadata = response.replicaMetadata();
```
Comment (frankgh) [inline cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/bulkwriter/token/TokenRangeMapping.java]: unused?

File: cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/bulkwriter/token/TokenRangeMapping.java [hunk 4]
```diff
@@ -22,55 +22,131 @@
...
-    private final transient RangeMap<BigInteger, List<Instance>> replicasByTokenRange;
-    private final transient Multimap<Instance, Range<BigInteger>> tokenRangeMap;
-    private final transient Map<String, Set<String>> writeReplicasByDC;
-    private final transient Map<String, Set<String>> pendingReplicasByDC;
-    private final transient List<ReplicaMetadata> replicaMetadata;
+    private final transient Set<I> allInstances;
+    private final transient RangeMap<BigInteger, List<I>> replicasByTokenRange;
+    private final transient Multimap<I, Range<BigInteger>> tokenRangeMap;
+    private final transient Map<String, Set<I>> writeReplicasByDC;
+    private final transient Map<String, Set<I>> pendingReplicasByDC;
```
Comment (frankgh) [inline cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/bulkwriter/token/TokenRangeMapping.java]: I don't think it's worth keeping the replicas by DC, we always end up flattening. Why not just keep the set instead?

**Review by frankgh — COMMENTED** [review]
Looks good in general, I've added some comments.

File: scripts/build-dtest-jars.sh (hunk 1 above)
Comment (yifan-c) [reply scripts/build-dtest-jars.sh:121]: discussed offline. Source code is no longer needed once dtest.jar is built. We are good with file deletion.

**Review by frankgh — APPROVED** [review]
+1

**Pull request merged**